snuba/datasets/__init__.py

```diff
@@ -1,85 +1,64 @@
-            default_topic: str,
+            table_writer: Optional[TableWriter],
             default_replacement_topic: Optional[str] = None,
             default_commit_log_topic: Optional[str] = None):
         self.__dataset_schemas = dataset_schemas
-        self.__processor = processor
-        self.__default_topic = default_topic
+        self.__table_writer = table_writer
         self.__default_replacement_topic = default_replacement_topic
         self.__default_commit_log_topic = default_commit_log_topic
 
     def get_dataset_schemas(self) -> DatasetSchemas:
+        """
+        Returns the collections of schemas for DDL operations and for
+        query.
+        See TableWriter to get a write schema.
+        """
         return self.__dataset_schemas
 
-    def get_processor(self) -> MessageProcessor:
-        return self.__processor
-
-    def get_writer(self, options=None, table_name=None) -> BatchWriter:
-        from snuba import settings
-        from snuba.clickhouse.http import HTTPBatchWriter
-
-        def default(value):
-            if isinstance(value, datetime):
-                return value.strftime(DATETIME_FORMAT)
-            else:
-                raise TypeError
-
-        return HTTPBatchWriter(
-            self.get_dataset_schemas().get_write_schema_enforce(),
-            settings.CLICKHOUSE_HOST,
-            settings.CLICKHOUSE_HTTP_PORT,
-            lambda row: json.dumps(row, default=default).encode("utf-8"),
-            options,
-            table_name,
-        )
-
-    def get_bulk_writer(self, options=None, table_name=None) -> BatchWriter:
+    def can_write(self) -> bool:
+        """
+        Returns True if this dataset has write capabilities
         """
-        This is a stripped down verison of the writer designed
-        for better performance when loading data in bulk.
+        return self.__table_writer is not None
+
+    def get_table_writer(self) -> TableWriter:
         """
-        # TODO: Consider using rapidjson to encode everywhere
-        # once we will be confident it is reliable enough.
-
-        from snuba import settings
-        from snuba.clickhouse.http import HTTPBatchWriter
-
-        return HTTPBatchWriter(
-            self.get_dataset_schemas().get_write_schema_enforce(),
-            settings.CLICKHOUSE_HOST,
-            settings.CLICKHOUSE_HTTP_PORT,
-            lambda row: rapidjson.dumps(row).encode("utf-8"),
-            options,
-            table_name,
-            chunk_size=settings.BULK_CLICKHOUSE_BUFFER,
-        )
+        Returns the TableWriter or throws if the dataaset is a readonly one.
+
+        Once we will have a full TableStorage implementation this method will
+        disappear since we will have a table storage factory that will return
+        only writable ones, scripts will depend on table storage instead of
+        going through datasets.
+        """
+        table_writer = self.__table_writer
+        assert \
```

Comment:
This is not 100% elegant. But at least it is just one step away from the dataset_factory and it will disappear with TableStorage.
Unfortunately I cannot solve this with inheritance (have a get_writable_dataset factory) because of TimeSeriesDataset which introduces an orthogonal dimension in the hierarchy.
I tried to put the write feature in a separate class and using multiple inheritance, but the result was not much better. I would have had to reimplement this method in all classes that extend the write class, and the factory would have to return such interface thus hiding the read features.

Reply:
I think this is alright for now — the `TimeSeriesDataset` should be going away relatively soon, and replaced by query extensions (and however we end up structuring the splitter), right?

I was also thinking that it might make sense to return an `Optional` from this and remove the assert so that we could rely on the type checker to tell us when we are doing something weird — but I guess almost everything we're doing right now is going to be weird (since the primary datasets are read/write) so maybe this isn't worth the trouble.

Reply:
Will see how it looks like with Optional.
I originally added the assertion because all the use cases (except for one in bootstrap) would turn into 
```
writer = get_table_writer()
assert writer is not None
```
Since all those case should break if we get there with a read only dataset